chain/traversal.go

```diff
@@ -0,0 +1,100 @@
+package chain
+
+import (
+	"context"
+	"gx/ipfs/QmR8BauakNcBa3RbE4nbQu76PDiJgoQgz8AJdhJuiU4TAw/go-cid"
+
+	"github.com/filecoin-project/go-filecoin/types"
+)
+
+// BlockProvider provides blocks. This is a subset of the ReadStore interface.
+type BlockProvider interface {
+	GetBlock(ctx context.Context, cid cid.Cid) (*types.Block, error)
+}
+
+// GetParentTipSet returns the parent tipset of a tipset.
+// The result is empty if the tipset has no parents (including if it is empty itself)
+func GetParentTipSet(ctx context.Context, store BlockProvider, ts types.TipSet) (types.TipSet, error) {
```

Comment:
This is moved from `message_queue.go`, the only change being the store parameter changing to this narrow interface in place of the concrete `hamt.CborIpldStore`.